File: python/encryption/MethodAES.py
import base64
import hashlib

import Crypto
from Crypto import AES
from Crypto import Counter


class MethodAES:

    def __init__(self, key, iv=None):
        self.key = key
        self.initial_vector = iv

    """ MODE ECB """
    def encrypt_ecb(self, text):
        key = MethodAES.key(self.key)
        text = MethodAES.pad(text)
        aes = AES.new(key, AES.MODE_ECB)
        return MethodAES.encrypt_and_encode(aes, text)

    def decrypt_ecb(self, text):
        key = MethodAES.key(self.key)
        text = base64.b64decode(text)
        aes = AES.new(key, AES.MODE_ECB)
        decrypted = aes.decrypt(text)
        decrypted = remove_padding(decrypted)
        return decrypted

    """ MODE CBC """
    def encrypt_cbc(self, text, iv=None):
        if iv is None:
            iv = AIOCrypto.Random.new().read(AES.block_size)

        key = MethodAES.key(self.key)
        text = MethodAES.pad(text)
        aes = AES.new(key, AES.MODE_CBC, iv)

        encrypted = aes.encrypt(text)

        return base64.b64encode(iv)[0:22] + base64.b64encode(encrypted)

    def decrypt_cbc(self, text):
        key = MethodAES.key(self.key)
        iv = base64.b64decode(text[0:22] + "==")
        text = base64.b64decode(text[22:len(text)])
        text = text.replace(iv, "")
        aes = AES.new(key, AES.MODE_CBC, iv)
        decrypted = aes.decrypt(text)
        decrypted = remove_padding(decrypted)
        return decrypted

    """ MODE CTR """
    def encrypt_ctr(self, text):
        key = MethodAES.key(self.key)
        text = MethodAES.pad(text)

        iv = Counter.new(128)
        aes = AES.new(key, AES.MODE_CTR, counter=iv)

        return MethodAES.encrypt_and_encode(aes, text)

    def decrypt_ctr(self, text):
        key = MethodAES.key(self.key)
        text = base64.b64decode(text)
        iv = Counter.new(128)
        aes = AES.new(key, AES.MODE_CTR, counter=iv)
        decrypted = aes.decrypt(text)
        decrypted = remove_padding(decrypted)
        return decrypted

    """ MODE CFB """
    def encrypt_cfb(self, text, iv=None):
        if iv is None:
            iv = AIOCrypto.Random.new().read(AES.block_size)

        key = MethodAES.key(self.key)
        text = MethodAES.pad(text)

        aes = AES.new(key, AES.MODE_CFB, iv)
        encrypted = aes.encrypt(text)
        return base64.b64encode(iv + encrypted)

    def decrypt_cfb(self, text):
        key = MethodAES.key(self.key)
        text = base64.b64decode(text)
        iv = text[0:AES.block_size]
        text = text.replace(iv, "")
        aes = AES.new(key, AES.MODE_CFB, iv)
        decrypted = aes.decrypt(text)
        decrypted = remove_padding(decrypted)
        return decrypted

    """ MODE OFB """
    def encrypt_ofb(self, text, iv=None):
        if iv is None:
            iv = AIOCrypto.Random.new().read(AES.block_size)

        key = MethodAES.key(self.key)
        text = MethodAES.pad(text)

        aes = AES.new(key, AES.MODE_OFB, iv)
        encrypted = aes.encrypt(text)
        return base64.b64encode(iv + encrypted)

    def decrypt_ofb(self, text):
        key = MethodAES.key(self.key)
        text = base64.b64decode(text)
        iv = text[0:AES.block_size]
        text = text.replace(iv, "")
        aes = AES.new(key, AES.MODE_OFB, iv)
        decrypted = aes.decrypt(text)
        decrypted = remove_padding(decrypted)
        return decrypted

    """ Static Methods """

    @staticmethod
    def key(key):
        return hashlib.sha256(key.encode('UTF-8')).digest()

    @staticmethod
    def pad(text):
        return pad(text, 16)

    @staticmethod
    def encode(text):
        return base64.b64encode(text)

    @staticmethod
    def encrypt_and_encode(aes, text):
        return MethodAES.encode(aes.encrypt(text))

    @staticmethod
    def test_all():
        line_break = "\n=============================MODE=============================\n"
        text = raw_input("Text to encrypt: ")  # "This is a confidential string"
        key = raw_input("Key for encryption: ")  # "key"
        iv = pad("hard code", 16)

        aes = MethodAES("test")
        print(line_break.replace("MODE", "Mode CBC With Random Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_cbc(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_cbc(encrypted))
            print("")

        quit()

        print("Given;\n\t-key\t\t\t\t'" + str(key) + "'\n\t-Text\t\t\t\t'" + str(
            text) + "'\n\t-Initial Vector\t\t'" + iv + "'")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode ECB"))
        for i in range(3):
            encrypted = aes.encrypt_ecb(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_ecb(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode CBC With Hardcoded Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_cbc(text, iv)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_cbc(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode CBC With Random Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_cbc(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_cbc(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode CTR"))
        for i in range(3):
            encrypted = aes.encrypt_ctr(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_ctr(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode CFB With Hardcoded Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_cfb(text, iv)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_cfb(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode CFB With Random Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_cfb(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_cfb(encrypted))
            print("")

        raw_input("Continue...")
        print(line_break.replace("MODE", "Mode OFB With Random Initial Vector"))
        for i in range(3):
            encrypted = aes.encrypt_ofb(text)
            print("Encrypted: " + encrypted)
            print("Encrypted Length: " + str(len(encrypted)))
            print("Decrypted: " + aes.decrypt_ofb(encrypted))
            print("")


def pad(text, buffer_size):
    number_of_blank_space = buffer_size - len(text) % buffer_size
    character_by_that_value = chr(number_of_blank_space)
    filler = number_of_blank_space * character_by_that_value
    return text + filler


def remove_padding(text):
    last_char = text[len(text) - 1:]
    number_of_blank_space = ord(last_char)
    length_of_text = len(text) - number_of_blank_space
    removed_filter = text[:length_of_text]
    return removed_filter

File: java/src/com/company/encryption/MethodAES.java
package com.company.encryption;

import com.company.hashing.SHA256;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.Charset;
import java.security.*;
import java.util.Arrays;
import java.util.Base64;
import java.util.Scanner;
import java.util.logging.Logger;

import static com.company.encryption.Encryption.getBytes;
import static java.lang.System.out;
import static java.util.logging.Level.SEVERE;

@SuppressWarnings("ALL")
public class MethodAES {

    private static final Logger LOGGER = Logger.getLogger(MethodAES.class.getSimpleName());

    private static Cipher ecb;
    private static Cipher cbc;
    static{
        try {
            ecb = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cbc = Cipher.getInstance("AES/CBC/PKCS5Padding");
        } catch (NoSuchAlgorithmException  | NoSuchPaddingException e) {
            LOGGER.log(SEVERE, "", e);
        }
    }


    private SecretKeySpec key;

    private String iv;

    private static final int BUFFER_SIZE = 16;

    private static final Charset CHARACTER_SET = Charset.forName("UTF-8");


    public MethodAES(String key, String iv){
        this.iv = pad(iv);
        byte[] keyBytes = new SHA256().hash(key);
        keyBytes = Arrays.copyOf(keyBytes, BUFFER_SIZE*2);
        this.key = new SecretKeySpec(keyBytes, "AES");
    }

    public MethodAES(String key){
        this.iv = null;
        byte[] keyBytes = new SHA256().hash(key);
        keyBytes = Arrays.copyOf(keyBytes, BUFFER_SIZE*2);
        this.key = new SecretKeySpec(keyBytes, "AES");
    }


    public String encryptECB(String text){
        try {
            ecb.init(Cipher.ENCRYPT_MODE, key);

            byte[] message = getBytes(text);
            message = ecb.doFinal(message);
            return Base64.getEncoder().encodeToString(message);
        } catch (InvalidKeyException | BadPaddingException | IllegalBlockSizeException e) {
            LOGGER.log(SEVERE, "", e);
        }
        return null;
    }

    public String decryptECB(String text){
        try {
            ecb.init(Cipher.DECRYPT_MODE, key);

            byte[] message = Base64.getDecoder().decode(getBytes(text));

            message = ecb.doFinal(message);
            return new String( message, CHARACTER_SET);
        } catch (InvalidKeyException | BadPaddingException | IllegalBlockSizeException e) {
            LOGGER.log(SEVERE, "", e);
        }
        return null;
    }


    public String encryptCBC(String text){
        try {
            byte[] initialVectorBytes = createIV();
            if( iv != null )
                initialVectorBytes = iv.getBytes();

            IvParameterSpec initialVector = new IvParameterSpec(initialVectorBytes);

            cbc.init(Cipher.ENCRYPT_MODE, key, initialVector);

            byte[] message = getBytes(text);
            message = cbc.doFinal(message);

            return Base64.getEncoder().encodeToString(initialVectorBytes).substring(0,22) + Base64.getEncoder().encodeToString(message);
        } catch (InvalidKeyException | BadPaddingException | InvalidAlgorithmParameterException | IllegalBlockSizeException e) {
            LOGGER.log(SEVERE, "", e);
        }
        return null;
    }


    public String decryptCBC(String text){
        try {
            IvParameterSpec initialVector = getIV(text);
            text = text.substring(22, text.length());
            byte[] message = Base64.getDecoder().decode( getBytes(text) );

            cbc.init(Cipher.DECRYPT_MODE, key, initialVector);

            message = cbc.doFinal(message);

            return new String( message, CHARACTER_SET);
        } catch (InvalidKeyException | BadPaddingException | InvalidAlgorithmParameterException | IllegalBlockSizeException e) {
            LOGGER.log(SEVERE, "", e);
        }
        return null;
    }


    public static String pad(String text){
        int numberOfBlankSpaces = BUFFER_SIZE - text.length() % BUFFER_SIZE;
        StringBuilder filler = new StringBuilder();
        for ( int i = 0; i < numberOfBlankSpaces; i++)
            filler.append(Character.toString((char) numberOfBlankSpaces));
        return text + filler;
    }

    public static String removePadding(String text ){
        char lastChar = text.charAt(text.length() - 1);
        int numberOfBlankSpace = (int) lastChar;
        int lengthOfText = text.length() - numberOfBlankSpace;
        return text.substring(0, lengthOfText );
    }

    private byte[] createIV() {
        final byte[] initialVector = new byte[16];
        final SecureRandom random = new SecureRandom();
        random.nextBytes(initialVector);
        return initialVector;
    }

    private IvParameterSpec getIV(String message) {
        String initialVectorStr = message.substring(0, 22);
        byte[] initialVector = Base64.getDecoder().decode(initialVectorStr + "==");
        return new IvParameterSpec(initialVector);
    }


    public static void testAll(){

        Scanner scanner = new Scanner(System.in);
        out.print("Key to encrypt with: ");
        String key = scanner.nextLine();
        out.print("String to encrypt: ");
        String text = scanner.nextLine();

        StringBuilder output = new StringBuilder("Given;\n\t-key\t\t\t\t'" + key + "'\n\t-Text\t\t\t\t'" + text + "'");
        MethodAES aes = new MethodAES(key);

        for( int i = 0; i < 3; i++){
            String encrypted = aes.encryptECB(text);
            output.append("Encrypted: ");
            output.append(encrypted);
            output.append("\n");
            output.append("Encrypted length: ");
            output.append(encrypted.length());
            output.append("\n");
            output.append("Decrypted: ");
            output.append(aes.decryptECB(encrypted));
            output.append("\n\n");
        }


        for( int i = 0; i < 3; i++){
            String encrypted = aes.encryptCBC(text);
            output.append("Encrypted: ");
            output.append(encrypted);
            output.append("\n");
            output.append("Encrypted length: ");
            output.append(encrypted.length());
            output.append("\n");
            output.append("Decrypted: ");
            output.append(aes.decryptCBC(encrypted));
            output.append("\n\n");
        }


        out.print(output);

    }

    public static void main(String[] args) {
        MethodAES.testAll();
    }
}
